### ISSUE TITLE
JRuby Canonicalizer Support for Nokogiri::XML::XML_C14N_1_0 and XML_C14N_1_1

### ISSUE THREAD

**Issue by dqdinh** [benoist/xmldsig#29]
Hi @benoist,

I'm interested in getting libsaml and xmldsig working in JRuby. I traced errors in libsaml back to various errors while running xmldsig under JRuby.

I was wondering if you can give me a little context on why you limited Nokogiri's canonicalization methods when in JRuby. Specifically, I'm trying to reason through the following lines below.

https://github.com/benoist/xmldsig/blob/master/lib/xmldsig/canonicalizer.rb#L21-L28

Thanks!

**Comment by dqdinh** [comment]
Actually, after more debugging, it seems like the issue is with nokogiri:

`[2] pry(#<RSpec::ExampleGroups::Xmldsig::VerifySignedDocuments::SpecFixturesSignedIdealTxt>)> signed_document.validate(certificate)
nokogiri.internals.c14n.CanonicalizationException: c14n.Canonicalizer.UnsupportedOperation
       	at nokogiri.internals.c14n.Canonicalizer20010315.engineCanonicalizeSubTree(Canonicalizer20010315.java:170)
       	at nokogiri.internals.c14n.Canonicalizer.canonicalizeSubtree(Canonicalizer.java:224)
       	at nokogiri.XmlDocument.canonicalize(XmlDocument.java:620)
       	at nokogiri.XmlDocument$INVOKER$i$0$3$canonicalize.call(XmlDocument$INVOKER$i$0$3$canonicalize.gen)
...
TypeError: no implicit conversion of nil into String
from /Users/ddinh/workspace/lookout/xmldsig/lib/xmldsig/reference.rb:54:in `calculate_digest_value'`

**Comment by benoist** [comment]
> Did you run into the same error? Is this why you limited Nokogiri's canonicalization methods when in JRuby?

Yeah I ran into the same errors and thats why I decided to limit the use of those methods.

**Comment by dqdinh** [comment]
I think I found a bug in the valid? method in Xmldsig::Signature. It's a three day weekend in the US so I'll confirm and submit a pull request next week. But here is a summary of what I found:

To validate a signed document, the signature(s) for that document have `valid?` called with the correct certificate. `valid?` is a method in Xmldsig::Signature that will run a series of checks, one of which is `validate_digest_values`.

`validate_digest_values` will take the existing signature digest value and validate it by comparing the value with a recalculated digest as specified by a reference transformation value. To recalculate the digest, `calculate_digest_value` in Xmldsig::Reference is called.

'calculate_digest_value' will find the part of the XML specified by the reference transform and calculate the digest. The error occurs in https://github.com/benoist/xmldsig/blob/master/lib/xmldsig/reference.rb#L54 where the the Canonicalizer is called but a Canonicalizer method is not defined. 

We see in https://github.com/benoist/xmldsig/blob/master/lib/xmldsig/canonicalizer.rb#L25 that the default method is Nokogiri::XML::XML_C14N_EXCLUSIVE_1_0 however Nokogiri defaults to XML_C14N_1_0 (http://www.rubydoc.info/github/sparklemotion/nokogiri/Nokogiri%2FXML%2FDocument%3Acanonicalize).

The result is running `Canonicalizer.new(transformed).canonicalize` returns nil (preceded by a giant jruby error trace). Then running `digest_method.digest` on nil reproduces the error I reported above:

```
TypeError: no implicit conversion of nil into String
from /Users/ddinh/workspace/lookout/xmldsig/lib/xmldsig/reference.rb:54:in `calculate_digest_value'
```
